actors/builtin/miner/expiration_queue.go

```diff
@@ -698,3 +712,122 @@ func groupSectorsByExpiration(sectorSize abi.SectorSize, sectors []*SectorOnChai
 	})
 	return sectorEpochSets
 }
+
+// Groups sectors into sets based on their Expiration field.
+// If sectors are not found in the expiration set corresponding to their expiration field
+// (i.e. they have been rescheduled) traverse expiration sets to for groups where these
+// sectors actually expire.
+// Groups will be returned in expiration order, earliest first.
+func (q *ExpirationQueue) findSectorsByExpiration(sectorSize abi.SectorSize, sectors []*SectorOnChainInfo) ([]sectorEpochSet, error) {
+	sectorsByExpiration := make(map[abi.ChainEpoch][]*SectorOnChainInfo)
+
+	for _, sector := range sectors {
+		qExpiration := q.quant.QuantizeUp(sector.Expiration)
+		sectorsByExpiration[qExpiration] = append(sectorsByExpiration[qExpiration], sector)
+	}
+
+	sectorEpochSets := make([]sectorEpochSet, 0, len(sectorsByExpiration))
+	allRescheduled := bitfield.New()
+
+	// This map iteration is non-deterministic but safe because we sort by epoch below.
+	for expiration, epochSectors := range sectorsByExpiration { //nolint:nomaprange // result is subsequently sorted
+		var es ExpirationSet
+		if err := q.mustGet(expiration, &es); err != nil {
+			return nil, err
+		}
+
+		sectorNumbers := []uint64{}
+		totalPower := NewPowerPairZero()
+		totalPledge := big.Zero()
+		for _, sector := range epochSectors {
+			if found, err := es.OnTimeSectors.IsSet(uint64(sector.SectorNumber)); err != nil {
+				return nil, err
+			} else if found {
+				sectorNumbers = append(sectorNumbers, uint64(sector.SectorNumber))
+				totalPower = totalPower.Add(PowerForSector(sectorSize, sector))
+				totalPledge = big.Add(totalPledge, sector.InitialPledge)
+			} else {
+				allRescheduled.Set(uint64(sector.SectorNumber))
+			}
+		}
+		if len(sectorNumbers) > 0 {
+			sectorEpochSets = append(sectorEpochSets, sectorEpochSet{
+				epoch:   expiration,
+				sectors: sectorNumbers,
+				power:   totalPower,
+				pledge:  totalPledge,
+				es:      &es,
+			})
+		}
+	}
+
+	// if sectors are missed because they have been rescheduled, traverse expiration sets to find and group them.
+	if empty, err := allRescheduled.IsEmpty(); err != nil {
+		return nil, err
+	} else if !empty {
+		err := q.traverse(func(epoch abi.ChainEpoch, es *ExpirationSet) (bool, error) {
+			intersect, err := bitfield.IntersectBitField(es.OnTimeSectors, allRescheduled)
+			if err != nil {
+				return false, err
+			}
```

Comment:
Blocking: this needs to check `es.EarlySectors` as well. Especially for the case where the sector has moved because it was rescheduled. I'm a bit surprised this escaped the great testing.

I think to hit this case we need to fault a sector (thus rescheduling it) and then terminate it (the other call path to this point, since it can't be faulted again).

Reply:
I'm with @wadeAlexC that I don't think we could exercise this code. If we stop assuming that grouped sectors can be a mix of faulty and non-faulty sectors, our downstream accounting gets a lot more complicated. We could add an assertion that no sectors in the set are early. I tried it an the tests still pass.

Reply:
Ignoring `es.EarlySectors` (now down in groupExpirationSet) still makes me quite uncomfortable. Adding an error check that no such sectors are being sought would help.